python/ray/autoscaler/v2/scheduler.py

```diff
@@ -223,7 +233,76 @@ def _compute_score(self) -> UtilizationScore:
         Returns:
             A utilization score for this node.
         """
-        pass
+
+        # Compute the number of resource types being scheduled.
+        num_matching_resource_types = 0
+        for req in self.sched_requests:
+            for resource_name in req.resources_bundle.keys():
+                if resource_name in self.total_resources:
+                    num_matching_resource_types += 1
+
+        # Compute the utilization rate for each resource type
+        util_by_resources = []
+        for k, v in self.total_resources.items():
+            if v == 0:
+                # Skip any zero values.
+                continue
+            if k in self.available_resources:
+                util = (v - self.available_resources.get(k, 0)) / v
+                assert util >= 0 and util <= 1, f"Invalid utilization: {util}"
+                util_by_resources.append(util)
+
+        # Prefer not to launch a GPU node if there aren't any GPU requirements in the
+        # resource bundle.
+        gpu_ok = True
+        if AUTOSCALER_CONSERVE_GPU_NODES:
+            # TODO: we should also generalize this optimization for accelerators.
+            # https://github.com/ray-project/ray/issues/43079
+            is_gpu_node = self.total_resources.get("GPU", 0) > 0
+            any_gpu_requests = any(
+                "GPU" in r.resources_bundle for r in self.sched_requests
+            )
+            if is_gpu_node and not any_gpu_requests:
+                gpu_ok = False
+
+        # Prioritize avoiding gpu nodes for non-gpu workloads first,
+        # then prioritize matching multiple resource types,
+        # then prioritize using all resources,
+        # then prioritize overall balance of multiple resources.
+        return (
```

Comment:
There is a `UtilizationScorer`?

Reply:
i think so - i believe it was the default scorer func. i guess it could be made into a. `UtilizationScorer` class.